src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs

```diff
@@ -21,24 +21,55 @@ internal sealed class HttpConnectionPools : IDisposable
         /// <summary>The pools, indexed by endpoint.</summary>
         private readonly ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool> _pools;
         /// <summary>Timer used to initiate cleaning of the pools.</summary>
-        private readonly Timer _cleaningTimer; // TODO #23149: Consider changing this to stop when _pools is empty.
+        private readonly Timer _cleaningTimer;
         /// <summary>The maximum number of connections allowed per pool. <see cref="int.MaxValue"/> indicates unlimited.</summary>
         private readonly int _maxConnectionsPerServer;
+        /// <summary>
+        /// Cached value of whether or not the pool is empty in order to avoid the expensive
+        /// <see cref="ConcurrentDictionary{TKey,TValue}.IsEmpty"/> call.
+        /// </summary>
+        private bool _poolIsEmpty;
+        /// <summary>Object used to synchronize access to state in the pool.</summary>
+        private object SyncObj => _pools;
 
         /// <summary>Initializes the pools.</summary>
         /// <param name="maxConnectionsPerServer">The maximum number of connections allowed per pool. <see cref="int.MaxValue"/> indicates unlimited.</param>
         public HttpConnectionPools(int maxConnectionsPerServer)
         {
             _maxConnectionsPerServer = maxConnectionsPerServer;
             _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
-            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, CleanPoolTimeoutMilliseconds, CleanPoolTimeoutMilliseconds);
+            // Start out with the timer not running, since we have no pools.
+            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, Timeout.Infinite, Timeout.Infinite);
+            _poolIsEmpty = true;
         }
 
         /// <summary>Gets a pool for the specified endpoint, adding one if none existed.</summary>
         /// <param name="key">The endpoint for the pool.</param>
         /// <returns>The retrieved pool.</returns>
-        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key) =>
-            _pools.GetOrAdd(key, (k, s) => new HttpConnectionPool(s), _maxConnectionsPerServer);
+        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key)
+        {
+            HttpConnectionPool pool;
+            while (!_pools.TryGetValue(key, out pool))
+            {
+                pool = new HttpConnectionPool(_maxConnectionsPerServer);
+                if (_pools.TryAdd(key, pool))
+                {
+                    // If we currently don't have any pools, a new one will be added for this
+                    // connection. We should then start reaping stale pools.
+                    lock (SyncObj)
+                    {
+                        if (_poolIsEmpty)
```

Comment:
Could we instead name this _timerIsRunning and invert the polarity?  That's really what it's tracking.